### ISSUE TITLE
Tighten rules for "goto" with anonymous functions and local functions

### ISSUE THREAD

**Issue by jskeet** [dotnet/csharpstandard#930]
Originally reported in #926, I don't believe we currently handle the scoping of labels properly when it comes to anonymous functions and local functions.

I think we're already covered for "into":

> A label can be referenced from `goto` statements ([§13.10.4](statements.md#13104-the-goto-statement)) within the scope of the label.

> *Note*: This means that `goto` statements can transfer control within blocks and out of blocks, but never into blocks. *end note*

A local function declaring a label would have to have a block body, so I don't think you'd be able to goto a label within it.

Worth checking whether goto *outside* the local function (or anonymous function) is actually prohibited in the spec though. (It clearly shouldn't be allowed.)

```csharp
class Program
{
    static void Main()
    {
        LocalFunction();
    testlabel:

        Action action = () =>
        {
            if (DateTime.UtcNow.Second > 10)
            {
                goto testlabel;
            }
        };

        if (DateTime.UtcNow.Second < 5)
        {
            goto testlabel;
        }

        void LocalFunction()
        {
            if (DateTime.UtcNow.Second > 10)
            {
                goto testlabel;
            }
        }
    }
}
```

Same error for both: CS0159 No such label 'testlabel' within the scope of the goto statement

(Interestingly enough, VS *does* autocomplete testlabel, and renaming testlabel renames the targets of the goto...)

I suspect we *might* be able to fix this in 7.3, which currently states:

> Each *block* or *switch_block* creates a separate declaration space for labels. Names are introduced into this declaration space through *labeled_statement*s, and the names are referenced through *goto_statement*s. The ***label declaration space*** of a block includes any nested blocks. Thus, within a nested block it is not possible to declare a label with the same name as a label in an enclosing block.

Although declaring `testlabel` again within `LocalFunction` gives a different error:

> CS0158 The label 'testlabel' shadows another label by the same name in a contained scope

**Comment by KalleOlaviNiemitalo** [comment]
`goto` from an anonymous function to a label outside it is disallowed by [12.19.3 Anonymous function bodies](https://github.com/dotnet/csharpstandard/blob/d538391e2615696c873282af2dec65ed6493fb8b/standard/expressions.md#12193-anonymous-function-bodies):

> It is a compile-time error for the body to contain a `goto` statement, a `break` statement, or a `continue` statement whose target is outside the body or within the body of a contained anonymous function.

The equivalent restriction for local functions may be missing.

**Comment by jskeet** [comment]
Thanks for finding that, @KalleOlaviNiemitalo. I'll create a PR for the equivalent for local functions. Hopefully this is nice and straightforward, in that case. (Feels a bit of a shame for that not to be in the goto statement itself, but never mind.)

**Comment by KalleOlaviNiemitalo** [comment]
In C# 8, one could perhaps add a general rule against `goto`, `break`, and `continue`  from a function to a different function -- regardless of whether the functions are members, local functions, or anonymous functions.

This might be complicated by C# 9 top-level statements, though.

**Comment by jskeet** [comment]
Possibly. It wouldn't hurt to add that redundantly to 13.10.4, keeping the restrictions in local/anonymous functions too.

**Comment by jskeet** [comment]
Fixed by #929.
